### ISSUE TITLE
Allow overlap of future transactions for forecasts

### ISSUE THREAD

**Issue by makdisse** [simonmichael/hledger#1236]
### Context

The journal documentation about [periodic transactions says](https://hledger.org/journal.html#periodic-transactions): `Forecasted transactions will begin only after the last non-forecasted transaction’s date.` I believe it would be useful to remove this limitation.

In developing countries it is very common to purchase for things and pay in instalments. One can easily have multiple future transactions in his journal defined for the next 6-8 months. Currently if you do this to an account that account will be excluded from the forecast.

 ### Request

- Add a flag (eg: `--allow-forecast-overlap`) that will allow the forecast to overlap dates where there are existing future transactions in the journal file
- The command should be smart enough to avoid duplicated transactions by matching the date, accounts used and amount
  - Example: 

**Given the following journal file `test.ledger`**:

```
~ monthly to 2021-05
    Expenses:Travel   1 GBP
    Assets:Bank

~ monthly from 2020-05 to 2020-09
    Expenses:Other   1 GBP
    Assets:Bank

2020-04-01 Spend travel
    Expenses:Travel   1 GBP
    Assets:Bank

2020-05-01 Spend travel
    Expenses:Travel   1 GBP
    Assets:Bank

2020-06-01 Spend other
    Expenses:Other   1 GBP
    Assets:Bank
```

**When I run the command**: `hledger bal -b 2020-03 -e 2021-03 --forecast -M -f test.ledger --allow-forecast-overlap`
**I will see the following output**:

```
Balance changes in 2020-04-01-2021-02-28:

                 ||    Apr     May     Jun     Jul     Aug     Sep     Oct     Nov     Dec     Jan     Feb
=================++========================================================================================
 Assets:Bank     || -1 GBP  -2 GBP  -2 GBP  -2 GBP  -2 GBP  -1 GBP  -1 GBP  -1 GBP  -1 GBP  -1 GBP  -1 GBP
 Expenses:Other  ||      0   1 GBP   1 GBP   1 GBP   1 GBP       0       0       0       0       0       0
 Expenses:Travel ||  1 GBP   1 GBP   1 GBP   1 GBP   1 GBP   1 GBP   1 GBP   1 GBP   1 GBP   1 GBP   1 GBP
-----------------++----------------------------------------------------------------------------------------
                 ||      0       0       0       0       0       0       0       0       0       0       0
```
- The feature should also allow for the following parameters:
  - `--begin-forecast` sets forecast start date, overriding what's in the periodic rule
  - `--end-forecast` sets forecast end date
  - These parameters will allow the user to constrain the forecasted transactions to a sub-period within the report

**Comment by adept** [comment]
This seems like an easy enough thing to add, at the expense of the proliferation of feature-specific commandline flags. OTOH, i don't see any other flag candidates that could be "conscripted into service' here.

@simonmichael , how opposed are you to extra flags here?

**Comment by simonmichael** [comment]
I think I was suggesting the new flag in this case, at least the begin one, in a related discussion, I have forgotten where. 

The proposed feature may have edge cases and unexpected usability issues, it’ll need some testing to find out.

**Comment by adept** [comment]
I think the other feature is #835

**Comment by simonmichael** [comment]
I see you suggested it first maybe, thanks for the linkage.

**Comment by simonmichael** [comment]
PS @adept, I think the "start from last recorded transaction" rule came from Ledger, right ? Or did you choose that ?

**Comment by adept** [comment]
I honestly don't remember. It might've been from hledger, but it is just as likely to be a sensible default chosen from scratch (i am saying this because i tried to come up with a different default just now and was unable to)

**Comment by simonmichael** [comment]
Update: @adept has added an optional period expression argument to `--forecast` in #1250, allowing an overlapping/past forecast period to be specified. Latest docs: https://hledger.org/journal.html#forecasting-with-periodic-transactions . @makdisse this covers the first part of your request.

**Comment by simonmichael** [comment]
Bounty payment received for this and #1234. @adept contributed this to the project (and I have applied it to #1177). @makdisse, thanks for supporting the project!

**Comment by makdisse** [comment]
@adept is it in your plans to deliver the second part of the request (ie: deduplication)?

**Comment by simonmichael** [comment]
This seems like a bug in --forecast=PERIOD, reported in chat by @makdisse:

```
~ monthly from next month
  (a)  1
```

Why do these start in februrary ? 

$ hledger --debug=2 reg --forecast=2020-2022
forecastspan flag: Just DateSpan 2020 - 01 - 01. . 2021 - 12 - 31
forecastspan: DateSpan 2020 - 01 - 01. . 2021 - 12 - 31
2021-02-01                                 (a)                                        1             1
2021-03-01                                 (a)                                        1             2
2021-04-01                                 (a)                                        1             3
2021-05-01                                 (a)                                        1             4
2021-06-01                                 (a)                                        1             5
2021-07-01                                 (a)                                        1             6
2021-08-01                                 (a)                                        1             7
2021-09-01                                 (a)                                        1             8
2021-10-01                                 (a)                                        1             9
2021-11-01                                 (a)                                        1            10
2021-12-01                                 (a)                                        1            11
```

**Comment by simonmichael** [comment]
Getting clear on the history here - I believe: 
- @makdisse offered a bounty for two items
- Item 1 "allow the forecast to overlap dates where there are existing future transactions in the journal file" was implemented by @adept
- @makdisse awarded the full bounty
- @adept contributed it back to the project
- Item 2 "the command should be smart enough to avoid duplicated transactions by matching the date, accounts used and amount" has not been implemented.

So the bounty was paid. Normally I'd leave the bounty label on this issue for easy finding. But a wish remains open. @makdisse do you want to keep this issue open ?

**Comment by simonmichael** [comment]
And I've opened #1845 to discuss that last problem report (the "~ monthly from next month" rule).

**Comment by simonmichael** [comment]
Closing this one. If someone wants to work on or fund the other wish (automatically deduplicate forecasted txns with existing ones), let it be a new issue.
